In [1]:
from transformers import pipeline

# BERT for fill-mask (its actual pretraining task)
bert_fill = pipeline('fill-mask', model='bert-base-uncased')

# GPT-2 for text generation
gpt2_gen = pipeline('text-generation', model='gpt2')

# BERT predicts masked token
bert_result = bert_fill("The animal didn't cross the street because [MASK] was too tired.")
print("BERT fill-mask predictions:")
for r in bert_result[:3]:
    print(f"  {r['token_str']}: {r['score']:.4f}")

# GPT-2 generates continuation
gpt2_result = gpt2_gen("The animal didn't cross the street because", 
                        max_new_tokens=20, 
                        num_return_sequences=1)
print("\nGPT-2 continuation:")
print(gpt2_result[0]['generated_text'])

/opt/miniconda3/lib/python3.13/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
Loading weights: 100%|██████████████████████████████████████████████████████████| 202/202 [00:00<00:00, 2110.35it/s]
BertForMaskedLM LOAD REPORT from: bert-base-uncased
Key                         | Status     |  | 
----------------------------+------------+--+-
cls.seq_relationship.bias   | UNEXPECTED |  | 
cls.seq_relationship.weight | UNEXPECTED |  | 
bert.pooler.dense.weight    | UNEXPECTED |  | 
bert.pooler.dense.bias      | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
Loading weights: 100%|██████████████████████████████████████████████████████████| 148/148 [00:00<00:00, 7197.68it/s]
Passing `generation_config` together with generation-related 

BERT fill-mask predictions:
  it: 0.7556
  he: 0.2109
  she: 0.0216

GPT-2 continuation:
The animal didn't cross the street because we know the animal must have been there to get it, but we do know that the animal was


## Summary

**What I built:** Hands-on comparison of BERT and GPT-2 using 
HuggingFace pipelines — demonstrated understanding vs generation.

**Key results:**
- BERT fill-mask: predicted "it" with 75.56% confidence (correct)
- GPT-2 generation: grammatically correct but semantically wandering

**BERT vs GPT architecture:**
- BERT: encoder, bidirectional attention, sees full context both ways
  - Pretrained with Masked Language Modeling
  - Used for: classification, embeddings, Q&A, NER
  - Can't generate: trained on full context, not next-token prediction
- GPT: decoder, causal/masked attention, sees only previous tokens
  - Pretrained with next-token prediction
  - Used for: text generation, code completion, chat
  - Can't fill-mask: never trained to use future context

**Why GPT-2 generation is incoherent:**
- Small model (117M parameters)
- No instruction fine-tuning or RLHF
- Raw next-token prediction without alignment
- Modern GPT-4/Claude add RLHF on top

**Key insight:** Architecture follows task. Understanding needs 
full context (bidirectional). Generation needs left-only context 
(causal) to avoid cheating during training. You can't swap them.

**Next:** Week 6 review — Day 42